sympy/core/basic.py

```diff
@@ -1525,6 +1525,52 @@ def match(self, pattern, old=False):
         pattern = sympify(pattern)
         return pattern.matches(self, old=old)
 
+    def partition(self, var):
+        """
+        Return a `dict` of subexpressions which contains `var`.
+        Subexpressions which doesn't contain `var` are considered to be constant.
+
+        Examples
+        ========
+
+        >>> from sympy import exp, sqrt, cos, sin
+        >>> from sympy.abc import x, a, b
+        >>> expr = a + b + x
+        >>> expr.partition(x)
+        {'Constant': [a, b], 'Symbol': [x]}
+        >>> expr = cos(a)*b*sin(x)
+        >>> expr.partition(x)
+        {'Constant': [b, cos(a)], 'sin': [sin(x)]}
+
+        When the expression head is not commutative('Add' and 'Mul' type
+        epressions), the same expression is returned in form of dict.
+
+        >>> exp(x).partition(x)
+        {'Constant': [], 'exp': [exp(x)]}
+        >>> exp(a).partition(x)
+        {'Constant': [exp(a)]}
+
+        """
+        leaves = {}
+        leaves['Constant'] = []
+
+        if self.__class__.__name__  not in ['Mul', 'Add']:
```

Comment:
```python
from .add import Add
from .mul import Mul
if not isinstance(self, [Mul, Add]):
```

This would care for a possible inheritance of Mul and Add.